diff -up nfs-utils-lib-1.1.5/libnfsidmap/umich_ldap.c.orig nfs-utils-lib-1.1.5/libnfsidmap/umich_ldap.c
--- nfs-utils-lib-1.1.5/libnfsidmap/umich_ldap.c.orig	2011-08-23 10:09:09.408418000 -0400
+++ nfs-utils-lib-1.1.5/libnfsidmap/umich_ldap.c	2011-08-23 10:10:06.200099000 -0400
@@ -32,6 +32,10 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
 #ifdef ENABLE_LDAP
 
 #include <sys/types.h>
