diff -up nfs-utils-lib-1.1.5/libnfsidmap/libnfsidmap.c.orig nfs-utils-lib-1.1.5/libnfsidmap/libnfsidmap.c
--- nfs-utils-lib-1.1.5/libnfsidmap/libnfsidmap.c.orig	2014-04-30 10:19:28.000000000 -0400
+++ nfs-utils-lib-1.1.5/libnfsidmap/libnfsidmap.c	2014-04-30 10:49:05.000000000 -0400
@@ -99,8 +99,12 @@ static char * toupper_str(char *s)
 static int id_as_chars(char *name, int *id)
 {
 	long int value = strtol(name, NULL, 10);
-	if (value == 0)
-		return 0;
+
+	if (value == 0) {
+		/* check for zero id values */
+		if (strcmp(name, "0") != 0)
+			return 0;
+	}
 	*id = (int)value;
 	return 1;
 }
